### PR TITLE
Feat/fly 2543

### PR DESCRIPTION
**Pull request by Hakob23**
## What

Add `PegOutEscrow` implementing the frozen `IPegOutEscrow` request/cancel path and getters against `FlyoverConfigurations`: fee split (no `gasFee`), dust change vs PegOut `dustThreshold`, `refundAddress(0)` → `msg.sender`, quote snapshot at request, and cancel refund while `REQUESTED`. Stub `claimPegOut` / `refundOnNoClaim` / `onSettlement`. No interface changes.

## Why

Commit-first peg-out needs an on-chain escrow so the user’s RBTC deposit is the only commitment before an LP claims. This PR lands request + cancel (and discovery getters) on top of FLY-2542 configs; claim/settlement stay stubbed for a follow-up.

## Type of Change

- [ ] Bug fix (non-breaking change)
- [x] New feature (non-breaking change)
- [ ] Breaking change
- [ ] Refactor (no intended behavior change)
- [ ] Security hardening
- [ ] Tests only
- [ ] Build/CI/Tooling
- [ ] Documentation

## Affected Areas

- [ ] PegIn flow (`registerPegIn`, `callForUser`, quote validation)
- [x] PegOut flow (`depositPegout`, `refundPegOut`, `refundUserPegOut`)
- [ ] Liquidity Provider lifecycle (register/update/status/resign/collateral/withdraw)
- [ ] Pause/operational controls
- [ ] Quote hashing/signature logic
- [ ] Bitcoin tx / PMT / merkle validation
- [ ] Deployment or upgrade scripts / Makefile targets
- [ ] CI workflows (`ci`, `fuzz`, `formal`, `slither`, `codeql`)
- [ ] Docs

## Risk & Security Review

- [x] Access control and authorization paths reviewed
- [x] Reentrancy and external call paths reviewed
- [x] Storage layout / upgrade safety reviewed (if upgradeable code changed)
- [x] Time/block/confirmation assumptions reviewed
- [ ] BTC tx output/index assumptions reviewed (if pegout validation touched)
- [x] No secrets or sensitive values added

## Test Plan

List what you ran locally and the outcome:

- [x] `npm run lint:sol` (solhint on `src/PegOutEscrow.sol`)
- [ ] `npm run lint:ts`
- [x] `npm run compile` / `forge build`
- [x] `forge test --match-path 'test/pegout-escrow/*'`
- [ ] `npm test`
- [ ] `npm run test:fuzz` (if logic/state transitions changed)
- [ ] `npm run test:invariant` (if invariant-sensitive logic changed)
- [ ] `npm run test:formal` (if formal-verification-sensitive logic changed)
- [ ] `npm run test:integration` (if integration paths changed)
- [ ] `npm run test:e2e` (if deployment/ownership flows changed)

Additional notes on test coverage, edge cases, and any tests intentionally skipped:

Unit coverage in `test/pegout-escrow/PegOutEscrow.t.sol` for request/cancel AC (happy path, zero-refund mapping, serviceability, dust refund vs fold, cancel auth/state, stubs). Claim/settlement intentionally not implemented.

## Deployment & Ops Impact

- [ ] No deployment impact
- [x] Requires deployment
- [ ] Requires upgrade
- [ ] Env/config changes required (`.env`, network RPCs, verification, etc.)
- [ ] Runbook/update instructions added to PR description

New upgradeable escrow contract; wire `initialize(admin, delay, pauseRegistry, pegOut, collateral, configurations)` after configs are live. Deploy scripts not included in this PR.

## Related Issues

- Jira: [FLY-2543](https://rsklabs.atlassian.net/browse/FLY-2543)
- Depends on: feat/FLY-2542 (peg-out FlyoverConfigurations impl)

## Reviewer Notes

- Frozen rules vs PoC: no `gasFee` in fee split; `refundAddress(0)` → `msg.sender`; quote `gasFee` / `depositConfirmations` snapshotted as `0` (not on frozen config).
- Dust read via narrow `IDustThreshold` (not on `IPegOut`); change emit is impl-only `EscrowPegOutChangePaid` (not on frozen escrow ABI).
- `claimPegOut` / `refundOnNoClaim` / `onSettlement` revert `PegOutPathNotImplemented` (TODOs marked).

## Screenshots / Logs (if applicable)

N/A

### CODE REVIEW

File: src/PegOutEscrow.sol [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+    IPegOutEscrow
+{
+    /// @custom:storage-location erc7201:rsk.flyover.PegOutEscrow
+    struct PegOutEscrowStorage {
+        IPegOut pegOutContract;
+        ICollateralManagement collateralManagement;
+        IFlyoverConfigurations configurations;
+        mapping(bytes32 => Quotes.PegOutQuote) quotes;
+        mapping(bytes32 => EscrowedPegOutState) state;
+        mapping(uint256 => bytes32) idByNonce;
```
Comment (Freshenext) [inline src/PegOutEscrow.sol]: idByNonce - what is this

File: src/PegOutEscrow.sol [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+        uint256 requestCount;
```
Comment (Freshenext) [inline src/PegOutEscrow.sol:42]: requestCount - what is this

File: src/PegOutEscrow.sol [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+/// PegOutContract. Fee / deadline parameters come from {IFlyoverConfigurations}.
+/// @author Rootstock Labs
+contract PegOutEscrow is
+    AccessControlDefaultAdminRulesUpgradeable,
+    EmergencyPause,
+    ReentrancyGuard,
+    IPegOutEscrow
+{
+    /// @custom:storage-location erc7201:rsk.flyover.PegOutEscrow
+    struct PegOutEscrowStorage {
```
Comment (Freshenext) [inline src/PegOutEscrow.sol:28]: describe each field as a comment here so that we know what each are

File: src/PegOutEscrow.sol [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+            depositConfirmations: 0,
+            transferConfirmations: uint16(confirmations),
+            depositAddress: destinationAddress,
+            btcRefundAddress: "",
+            lpBtcAddress: ""
+        });
+    }
+
+    /// @dev Frozen inverse fee split + dust-change against PegOutContract.dustThreshold.
+    function _splitValue(
```
Comment (Freshenext) [inline src/PegOutEscrow.sol:314]: Could we have a comment explaining what this specifically does?

Can the variables be named more detailed, for example I see a "required" -> required what?

I had to go down and see the amount - callFee < required to know about this - I think we could name it more meaningful.

**Review by Freshenext — COMMENTED** [review]
Very formidable PR! A few comments

File: src/PegOutEscrow.sol [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+import {IFlyoverConfigurations} from "./interfaces/IFlyoverConfigurations.sol";
+import {IPauseRegistry} from "./interfaces/IPauseRegistry.sol";
+import {IPegOut} from "./interfaces/IPegOut.sol";
+import {IPegOutEscrow} from "./interfaces/IPegOutEscrow.sol";
+import {Flyover} from "./libraries/Flyover.sol";
+import {Quotes} from "./libraries/Quotes.sol";
+
+/// @dev Narrow read of PegOutContract's public dust threshold (avoid concrete import).
+/// TODO: drop this if dustThreshold is added to {IPegOut} (or another shared interface).
+interface IDustThreshold {
```
Comment (Luisfc68) [inline src/PegOutEscrow.sol]: I think if you need this as part of the interface you could add it to the pegout contract interface, I don't see any problem

File: src/PegOutEscrow.sol [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+        $.configurations = IFlyoverConfigurations(payable(configurations_));
+    }
+
+    /// @inheritdoc IPegOutEscrow
+    function requestPegOut(
+        bytes calldata destinationAddress,
+        address refundAddress
+    ) external payable override nonReentrant whenNotSoftPaused returns (bytes32 requestHash) {
+        if (destinationAddress.length == 0) revert InvalidDestination();
+        if (refundAddress == address(0)) {
```
Comment (Luisfc68) [inline src/PegOutEscrow.sol:127]: Following my comment in the interfaces PR: IMO we shouldn't allow the client to do this and make them explicitly say the refunded address

File: src/PegOutEscrow.sol (hunk 4 above)
Comment (Luisfc68) [inline src/PegOutEscrow.sol:314]: if this is private part of the contract why does it need the storage as parameter?

Comment (Luisfc68) [reply src/PegOutEscrow.sol:314]: If the answer is to not read it twice in the same function, its ok as long as you check it is actually cheaper

File: src/PegOutEscrow.sol [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+    }
+
+    function _payout(address to, uint256 amount) private {
+        (bool sent, bytes memory reason) = to.call{value: amount}("");
+        if (!sent) {
+            revert Flyover.PaymentFailed(to, amount, reason);
+        }
+    }
+
+    function _computeRequestHash(
```
Comment (Luisfc68) [inline src/PegOutEscrow.sol:251]: why this is different from a quote? the only diff between this and the quote is that the quote doens't have an LP assigned, so my question is, why the request hash can't be the quote with address(0) as LP? what is the advantage on having a separated formula?

File: src/PegOutEscrow.sol [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+
+        uint256 nonce = ++$.requestCount;
+        requestHash = _computeRequestHash(nonce, refundAddress, destinationAddress, amount, callFee);
+
+        $.state[requestHash] = EscrowedPegOutState.REQUESTED;
+        $.idByNonce[nonce] = requestHash;
+
+        uint256 confirmations = $.configurations.getRequiredPegOutBtcConfirmations(amount);
+        $.quotes[requestHash] =
+            _buildQuote($, cfg, refundAddress, destinationAddress, amount, callFee, nonce, confirmations);
```
Comment (Luisfc68) [inline src/PegOutEscrow.sol:149]: you're building the quote anyways thats another reason to reuse the partial quote as request hash

File: src/PegOutEscrow.sol [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+
+        emit PegOutRequested(requestHash, refundAddress, amount, destinationAddress);
+
+        if (change > 0) {
```
Comment (Luisfc68) [inline src/PegOutEscrow.sol]: shouldn't it be if change > dust?

File: src/PegOutEscrow.sol [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+        PegOutEscrowStorage storage $,
+        bytes32 requestHash,
+        EscrowedPegOutState finalState
+    ) private {
+        $.state[requestHash] = finalState;
+        delete $.quotes[requestHash];
+    }
+
+    function _payout(address to, uint256 amount) private {
+        (bool sent, bytes memory reason) = to.call{value: amount}("");
```
Comment (Luisfc68) [inline src/PegOutEscrow.sol]: change this to avoid the return bomb by making a safe call
```
        address target = quote.contractAddress;
        uint256 value = quote.value;
        uint256 gasLimit = quote.gasLimit;
        bytes memory data = quote.data;
        assembly {
            success := call(gasLimit, target, value, add(data, 0x20), mload(data), 0, 0)
        }
```

File: src/PegOutEscrow.sol (hunk 5 above)
Comment (Luisfc68) [reply src/PegOutEscrow.sol]: or move it to a separate file if you plan to reuse in other sources

**Review by Luisfc68 — CHANGES_REQUESTED** [review]

File: src/PegOutEscrow.sol (hunk 6 above)
Comment (Hakob23) [reply src/PegOutEscrow.sol:127]: The interface natspec freeze version explicitly says it, that's why I left it this way. we could stack it inside the future changes list.

File: src/PegOutEscrow.sol (hunk 4 above)
Comment (Hakob23) [reply src/PegOutEscrow.sol:314]: Passing the storage pointer avoids a second `_getStorage()` and keeps one `$` for `configurations / dust`

File: src/PegOutEscrow.sol (hunk 7 above)
Comment (Hakob23) [reply src/PegOutEscrow.sol:251]: At request time, no LP has claimed yet, so we don’t know who the LP will be. The stored quote still has to fill `lpRskAddress`, so we put `address(0)` there as a placeholder. That quote is only a snapshot of the terms — not the id.

The request id (requestHash) is a different, smaller hash from a fixed list of fields: chain, escrow, nonce, sender, refund address, destination, amount, fee, timestamp.

IMO we should keep them separate because the id must exist before an LP is chosen.
The freeze explicitly named it requestHash so nobody confuses it with the normal peg-out quote hash.

File: src/PegOutEscrow.sol (hunk 8 above)
Comment (Hakob23) [reply src/PegOutEscrow.sol:149]: As discusssed here: https://github.com/rsksmart/liquidity-bridge-contract/pull/529#discussion_r3736729892

File: src/PegOutEscrow.sol (hunk 10 above)
Comment (Hakob23) [reply src/PegOutEscrow.sol]: https://github.com/rsksmart/liquidity-bridge-contract/pull/529/commits/65e134469791588e8cca6234776bb06c0a9d58b2

File: src/PegOutEscrow.sol (hunk 9 above)
Comment (Hakob23) [reply src/PegOutEscrow.sol]: No, this `if (changeRefund > 0)` is only the payout guard after dust is already applied in `_splitValue`.

File: src/PegOutEscrow.sol (hunk 5 above)
Comment (Hakob23) [reply src/PegOutEscrow.sol]: https://github.com/rsksmart/liquidity-bridge-contract/pull/529/commits/78bf4962deb3991c3a4c16ab04d3121d4b14d6be

File: src/PegOutEscrow.sol (hunk 4 above)
Comment (Hakob23) [reply src/PegOutEscrow.sol:314]: https://github.com/rsksmart/liquidity-bridge-contract/pull/529/commits/ee7737e7c8be90017ad4c59b6d9e9c64a77e4e79

File: src/PegOutEscrow.sol (hunk 3 above)
Comment (Hakob23) [reply src/PegOutEscrow.sol:28]: https://github.com/rsksmart/liquidity-bridge-contract/pull/529/commits/ee7737e7c8be90017ad4c59b6d9e9c64a77e4e79

File: src/PegOutEscrow.sol (hunk 2 above)
Comment (Hakob23) [reply src/PegOutEscrow.sol:42]: /// How many requests have been minted; next request uses `++requestCount`.

This comment has been added to the contract https://github.com/rsksmart/liquidity-bridge-contract/pull/529/commits/ee7737e7c8be90017ad4c59b6d9e9c64a77e4e79

File: src/PegOutEscrow.sol (hunk 1 above)
Comment (Hakob23) [reply src/PegOutEscrow.sol]: /// requestHash for each 1-based(starting from 1 not 0) nonce; LPS rebuild after missed events

changed the name to `requestHashByNonce` for a better clarity.

https://github.com/rsksmart/liquidity-bridge-contract/pull/529/commits/ee7737e7c8be90017ad4c59b6d9e9c64a77e4e79

File: src/PegOutEscrow.sol (hunk 6 above)
Comment (Luisfc68) [reply src/PegOutEscrow.sol:127]: Yes please

File: src/PegOutEscrow.sol (hunk 7 above)
Comment (Luisfc68) [reply src/PegOutEscrow.sol:251]: you can still name it "requestHash" what I'm not sure is that we should use a separate formula, IMO we should maintain the quote hash, if you fix the lp address in the quote hash to address(0) you get the id that you want before the LP is chosen as you needed, doing it like this means maintaining 2 different calculations, if the quote is extended to (for example) include the address type of the destination, we'll need to remember to add it here, because the code won't break in that scenario, and if we forget we might end up in inconsistencies that can result in vulnerabilities

File: src/PegOutEscrow.sol [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+    function _splitValue(
+        PegOutEscrowStorage storage $,
+        uint256 value,
+        IFlyoverConfigurations.PegOutConfiguration memory cfg
+    ) private view returns (uint256 amount, uint256 callFee, uint256 changeRefund) {
+        // solhint-disable-next-line gas-strict-inequalities
+        if (value <= cfg.fixedFee) {
+            revert Flyover.InsufficientAmount(value, cfg.fixedFee + 1);
+        }
+        amount = ((value - cfg.fixedFee) * FEE_PERCENTAGE_DENOMINATOR)
```
Comment (Luisfc68) [inline src/PegOutEscrow.sol:323]: can you explain this? you're calculating the fee here to subtract it from the amount to add it again later? I'm not sure about this, if `calculatePegOutFee` wouldn't you be forced to change this too? perhaps this should be done in configurations and not in the escrow

File: src/PegOutEscrow.sol (hunk 5 above)
Comment (Hakob23) [reply src/PegOutEscrow.sol]: Follow-up kept settlement on PegOut's existing surface (`depositPegOut` / notify). See https://github.com/rsksmart/liquidity-bridge-contract/pull/549.

File: src/PegOutEscrow.sol (hunk 6 above)
Comment (Hakob23) [reply src/PegOutEscrow.sol]: Zero refund address is rejected; client must set it explicitly. https://github.com/rsksmart/liquidity-bridge-contract/pull/549

File: src/PegOutEscrow.sol (hunk 7 above)
Comment (Hakob23) [reply src/PegOutEscrow.sol]: Agreed — request hash is the incomplete quote hash (`lpRskAddress = 0`), not a custom preimage. https://github.com/rsksmart/liquidity-bridge-contract/pull/549

File: src/PegOutEscrow.sol (hunk 8 above)
Comment (Hakob23) [reply src/PegOutEscrow.sol]: Same as above — reuse the partial quote as the request id. https://github.com/rsksmart/liquidity-bridge-contract/pull/549

File: src/PegOutEscrow.sol (hunk 9 above)
Comment (Hakob23) [reply src/PegOutEscrow.sol]: Aligned with PegOut dust rule (refund when change ≥ dust). https://github.com/rsksmart/liquidity-bridge-contract/pull/549

File: src/PegOutEscrow.sol (hunk 10 above)
Comment (Hakob23) [reply src/PegOutEscrow.sol]: Escrow payouts use the safe low-level call pattern. https://github.com/rsksmart/liquidity-bridge-contract/pull/549

File: src/PegOutEscrow.sol (hunk 4 above)
Comment (Hakob23) [reply src/PegOutEscrow.sol]: Internal helper still takes storage for ERC-7201 access; happy to revisit style in a follow-up if you want it flattened.

File: src/PegOutEscrow.sol (hunk 11 above)
Comment (Hakob23) [reply src/PegOutEscrow.sol]: Fee split now goes through shared config helpers / `calculatePegOutFee` rather than duplicating fee math. https://github.com/rsksmart/liquidity-bridge-contract/pull/549
